components/sup/src/manager/service/pipe_hook_client.rs

```diff
@@ -0,0 +1,319 @@
+    hook_path:       PathBuf,
+    pipe_name:       String,
+    stdout_log_file: PathBuf,
+    stderr_log_file: PathBuf,
+}
+
+impl PipeHookClient {
+    pub fn new(hook_name: String,
+               hook_path: PathBuf,
+               stdout_log_file: PathBuf,
+               stderr_log_file: PathBuf)
+               -> Self {
+        let pipe_name = format!("{}-hook-{}", hook_name, Uuid::new_v4());
+        Self { hook_name,
+               hook_path,
+               pipe_name,
+               stdout_log_file,
+               stderr_log_file }
+    }
+
+    pub fn exec_hook<T>(&self,
+                        service_group: &str,
+                        pkg: &Pkg,
+                        svc_encrypted_password: Option<T>)
+                        -> Result<u32>
+        where T: ToString
+    {
+        // Always check to see if the powershell named pipe server is running.
+        // This should only be false the first time this function is called
+        // or if someone explicitly terminates the powershell process
+        if self.pipe_exists().is_err() {
+            debug!("Initiating Powershell named pipe server for {} {} hook",
+                   service_group, self.hook_name);
+            self.init_server(service_group, pkg, svc_encrypted_password)?;
+        }
+
+        let (mut pipe, poll) = self.connect()?;
+        debug!("connected to {} {} hook pipe",
+               service_group, self.hook_name);
+
+        // The powershell server takes a single byte as input which will be either
+        // 0 to shut down (see drop below) or 1 to run the hook
+        self.pipe_ready(&poll, Ready::writable())?;
+        pipe.write_all(&SIGNAL_EXEC_HOOK)?;
+
+        // Now we wait for the hook to run and the powershell service to
+        // send back the hook's exit code over the pipe
+        self.pipe_ready(&poll, Ready::readable())?;
+        let mut exit_buf = [0; 4];
+        pipe.read_exact(&mut exit_buf)?;
+        Ok(u32::from_ne_bytes(exit_buf))
+    }
+
+    fn pipe_ready(&self, poll: &Poll, readiness: Ready) -> Result<()> {
+        let mut events = Events::with_capacity(1024);
+        loop {
+            poll.poll(&mut events, None)?;
+            for event in &events {
+                if event.readiness().contains(readiness) {
+                    return Ok(());
+                }
+            }
+        }
+    }
+
+    fn abs_pipe_name(&self) -> String { format!("\\\\.\\pipe\\{}", self.pipe_name) }
+
+    fn connect(&self) -> Result<(NamedPipe, Poll)> {
+        let mut opts = OpenOptions::new();
+        opts.read(true)
+            .write(true)
+            .custom_flags(winbase::FILE_FLAG_OVERLAPPED);
+        let file = opts.open(self.abs_pipe_name())?;
+
+        let pipe = unsafe { NamedPipe::from_raw_handle(file.into_raw_handle()) };
+        let poll = Poll::new()?;
+        poll.register(&pipe, Token(0), Ready::all(), PollOpt::edge())?;
+        Ok((pipe, poll))
+    }
+
+    fn init_server<T>(&self,
```

Comment:
Along the lines of the above comment, we don't "expect" to call it multiple times but we need to be prepared to call it multiple times in the event the hosting `pwsh` process is terminated. I like keeping `connect` focused on connecting to an existing pipe. It also avoids needing to pass in the args that need to be forwarded to `init_service`.  For instance, the `quit` function connects if there is an existing pipe but there is no need to create a new pipe service just to quit.

I had originally implemented this using a durrable connection and the struct held an `Option<Pipe>` in a mutatable property. The prblem that introduced is that the `hook`'s `run` function is not setup to be mutable which it needs to be since it's pipe client can mutate. However changing the signature of `run` to use a `&mut self` proved to be pretty invasive.

I can definitely change the name of the `init_server` function though. Perhaps to `start_server`.